sdcm/utils/nemesis_utils/node_allocator.py

```diff
@@ -0,0 +1,247 @@
+
+            selected_node: 'BaseNode'
+            if allow_only_last_node_in_rack and rack is not None:
+                selected_node = candidate_nodes[-1]
+            else:
+                selected_node = random.choice(candidate_nodes)
+
+            self.active_nemeses_on_nodes[selected_node] = nemesis_name
+            selected_node.running_nemesis = nemesis_name
+            LOGGER.info("%s: selected target node %s. Marked it as running nemesis.",
+                        nemesis_name, selected_node.name)
+            return selected_node
+
+    def set_running_nemesis(self, node: 'BaseNode', nemesis_name: str) -> bool:
+        """
+        Sets running nemesis for the node.
+
+        This is for the cases when e.g. a new node is added by a nemesis, which is running on another target node.
+        """
+        with self.lock:
+            if node in self.active_nemeses_on_nodes:
+                LOGGER.error(
+                    "%s: Tried to set running nemesis for node %s, but it was already reserved by '%s' nemesis.",
+                    nemesis_name, node.name, self.active_nemeses_on_nodes[node])
+                return False
+
+            self.active_nemeses_on_nodes[node] = nemesis_name
+            node.running_nemesis = nemesis_name
+            LOGGER.info("%s: set running nemesis for node %s.", nemesis_name, node.name)
+            return True
+
+    def unset_running_nemesis(self, node: 'BaseNode', nemesis_name: str):
+        """Unsets running nemesis, making the node available for other nemeses."""
+        with self.lock:
+            if node in self.active_nemeses_on_nodes:
+                if self.active_nemeses_on_nodes[node] == nemesis_name:
+                    del self.active_nemeses_on_nodes[node]
+                    node.running_nemesis = None
+                    LOGGER.info("%s: unset running nemesis for node %s.", nemesis_name, node.name)
+                else:
+                    LOGGER.error("%s: tried to release node %s, but it was running another '%s' nemesis.",
+                                 nemesis_name, node.name, self.active_nemeses_on_nodes[node])
+            else:
+                LOGGER.error("%s: tried to release node %s, but it was not running any nemesis.",
+                             nemesis_name, node.name)
+
+    def unset_running_nemesis_from_all_nodes(self, nemesis_name: str):
+        """
+        Unsets the specified nemesis from all nodes that run it.
+
+        The primary use case is when a nemesis creates new node(s), which are supposed to stay in the cluster
+        after the nemesis is finished. To avoid stale records in the allocator, and for the subsequent nemeses to
+        be able to use the new node(s), we need to release them after disruption is finished.
+        """
+        with self.lock:
+            nodes_to_release = [node for node, owner in self.active_nemeses_on_nodes.items() if owner == nemesis_name]
```

Comment:
I'm afraid name of nemesis is enough, what happens if two nemesis threads are running the exact same nemesis ?